compiler/next/include/chpl/Queries/ContextDetail.h

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_QUERIES_CONTEXT_DETAIL_H
+#define CHPL_QUERIES_CONTEXT_DETAIL_H
+
+#include "chpl/AST/UniqueString.h"
+#include "chpl/AST/ErrorMessage.h"
+#include "chpl/Util/memory.h"
+
+#include <cstring>
+#include <unordered_map>
+#include <vector>
+
+/// \cond DO_NOT_DOCUMENT
+namespace chpl {
+namespace detail {
+
+struct UniqueStrEqual final {
+  bool operator()(const char* lhs, const char* rhs) const {
+    // pass the 2 metadata bytes
+    return strcmp(lhs, rhs) == 0;
+  }
+};
+
+struct UniqueStrHash final {
+  std::size_t operator()(const char* s) const {
+    // this hash is from StringHashFns in the old map.h
+    unsigned int h = 0;
+    while (*s) h = h * 27 + (unsigned char)*s++;
```

Comment:
Consider FNV-1a

Reply:
Can you provide a link or something? I'm not familiar with it.

Reply:
https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function Is actually pretty decent.

Reply:
I think I'd like to leave as future work changing these hash functions to FNV-1a. Not being familiar with either, it seems like something we should try to evaluate in some manner, in order to choose the better one. But that seems like something we should have as a separate effort.

Reply:
The link to my tuple code has FNV-1a implemented.